File: packages/cli/src/commands/suggest.ts
import { searchPrompts, semanticRerank, buildIndex, type SearchResult, type RankedResult } from "@jeffreysprompts/core/search";
import { type Prompt } from "@jeffreysprompts/core/prompts";
import chalk from "chalk";
import { shouldOutputJson } from "../lib/utils";
import { loadRegistry } from "../lib/registry-loader";

interface SuggestOptions {
  json?: boolean;
  limit?: number | string;
  semantic?: boolean;
}

interface SuggestOutput {
  task: string;
  suggestions: Array<{
    id: string;
    title: string;
    description: string;
    category: string;
    relevance: number;
    matchedFields: string[];
    tip: string;
  }>;
  total: number;
}

/**
 * jfp suggest <task> - Suggest prompts for a task
 *
 * Uses BM25 search to find relevant prompts for a task description.
 * With --semantic flag, applies MiniLM reranking for better semantic matching.
 * Falls back to hash-based embeddings if the model is unavailable.
 * Returns scored results with explanations of why they match.
 */
export async function suggestCommand(task: string, options: SuggestOptions) {
  const limit = options.limit !== undefined ? Number(options.limit) : 3;
  if (!Number.isFinite(limit) || limit <= 0) {
    if (shouldOutputJson(options)) {
      console.log(JSON.stringify({ error: "invalid_limit", message: "Provide a positive number for --limit." }));
    } else {
      console.error(chalk.red("Invalid --limit value. Provide a positive number."));
    }
    process.exit(1);
  }

  // Clamp limit to prevent OOM/performance issues
  const clampedLimit = Math.min(limit, 100);
  if (limit > 100 && !shouldOutputJson(options)) {
    console.warn(chalk.yellow(`Warning: Limit capped to 100 for performance.`));
  }

  if (!task.trim()) {
    if (shouldOutputJson(options)) {
      console.log(JSON.stringify({ error: "empty_task", message: "Please provide a task description" }));
    } else {
      console.error(chalk.red("Please provide a task description."));
      console.log(chalk.dim("Example: jfp suggest \"improve documentation for my API\""));
    }
    process.exit(1);
  }

  // Load registry dynamically (SWR pattern)
  const registry = await loadRegistry();
  const prompts = registry.prompts;
  
  // Build lookup map and search index
  const promptsMap = new Map(prompts.map((p) => [p.id, p]));
  const searchIndex = buildIndex(prompts);

  // Search using BM25 - get more results if semantic reranking will be applied
  const searchLimit = options.semantic ? Math.max(clampedLimit * 3, 10) : clampedLimit;
  let results = searchPrompts(task, { 
    limit: searchLimit,
    index: searchIndex,
    promptsMap: promptsMap
  });

  // Apply semantic reranking if requested
  if (options.semantic && results.length > 0) {
    if (!shouldOutputJson(options)) {
      console.log(chalk.dim("Applying semantic reranking..."));
    }

    // Convert to RankedResult format for semantic rerank
    const rankedResults: RankedResult[] = results.map((r) => ({
      id: r.prompt.id,
      score: r.score,
      text: `${r.prompt.title} ${r.prompt.description} ${r.prompt.tags.join(" ")}`,
    }));

    // Apply semantic reranking with hash fallback
    const reranked = await semanticRerank(task, rankedResults, {
      topN: searchLimit,
      fallback: "hash",
    });

    // Map back to SearchResult format, maintaining the reranked order
    // We use promptsMap for O(1) lookups
    results = reranked
      .map((ranked, idx) => {
        const prompt = promptsMap.get(ranked.id);
        if (!prompt) return null;
        
        // Find original result to preserve matchedFields if possible, or recreate basic result
        const original = results.find(r => r.prompt.id === ranked.id);
        
        return {
          prompt,
          score: ranked.score, // Use semantic score
          matchedFields: original?.matchedFields ?? [],
        };
      })
      .filter((r): r is SearchResult => r !== null)
      .slice(0, clampedLimit);
  } else {
    // Just trim to requested limit
    results = results.slice(0, clampedLimit);
  }

  if (shouldOutputJson(options)) {
    const output: SuggestOutput = {
      task,
      suggestions: results.map((r) => formatSuggestion(r, task, options.semantic)),
      total: results.length,
    };
    console.log(JSON.stringify(output, null, 2));
    return;
  }

  // Human-readable output
  if (results.length === 0) {
    console.log(chalk.yellow("No relevant prompts found for this task."));
    console.log(chalk.dim("Try a different task description or use 'jfp list' to see all prompts."));
    return;
  }

  console.log(chalk.bold.cyan(`Prompts for: "${task}"\n`));

  for (let i = 0; i < results.length; i++) {
    const { prompt, score, matchedFields } = results[i];
    // Semantic scores are 0-1 (need * 100), BM25 scores are typically 0-5 (need * 20)
    const relevancePercent = Math.min(100, Math.round(options.semantic ? score * 100 : score * 20));
    const relevanceBar = getRelevanceBar(relevancePercent);

    // Header: rank, title, relevance
    console.log(
      chalk.bold.white(`${i + 1}. ${prompt.title}`) +
      chalk.dim(` (${prompt.id})`)
    );

    // Relevance visualization
    console.log(
      chalk.dim("   Relevance: ") +
      relevanceBar +
      chalk.dim(` ${relevancePercent}%`)
    );

    // Category and description
    console.log(
      chalk.dim("   ") +
      chalk.green(`[${prompt.category}]`) +
      chalk.dim(" ") +
      prompt.description
    );

    // Why this prompt matches
    const tip = generateTip(prompt, task, matchedFields);
    console.log(chalk.dim("   → ") + chalk.italic(tip));

    console.log();
  }

  // Footer with usage hint
  console.log(chalk.dim("─".repeat(50)));
  console.log(chalk.dim("Use ") + chalk.cyan("jfp show <id>") + chalk.dim(" to view full prompt"));
  console.log(chalk.dim("Use ") + chalk.cyan("jfp copy <id>") + chalk.dim(" to copy to clipboard"));
}

/**
 * Generate a visual relevance bar
 */
function getRelevanceBar(percent: number): string {
  const filled = Math.round(percent / 10);
  const empty = 10 - filled;

  let color: (s: string) => string;
  if (percent >= 70) {
    color = chalk.green;
  } else if (percent >= 40) {
    color = chalk.yellow;
  } else {
    color = chalk.red;
  }

  return color("█".repeat(filled)) + chalk.dim("░".repeat(empty));
}

/**
 * Generate a contextual tip explaining why this prompt matches
 */
function generateTip(
  prompt: { title: string; category: string; tags: string[]; description: string },
  task: string,
  matchedFields: string[]
): string {
  const taskLower = task.toLowerCase();

  // Check for keyword matches to provide specific tips
  if (taskLower.includes("idea") || taskLower.includes("brainstorm") || taskLower.includes("improve")) {
    if (prompt.category === "ideation") {
      return "Great for generating and evaluating improvement ideas";
    }
  }

  if (taskLower.includes("doc") || taskLower.includes("readme") || taskLower.includes("documentation")) {
    if (prompt.category === "documentation") {
      return "Helps maintain clear, up-to-date documentation";
    }
  }

  if (taskLower.includes("cli") || taskLower.includes("tool") || taskLower.includes("automat")) {
    if (prompt.category === "automation") {
      return "Creates agent-friendly tools and automation";
    }
  }

  if (taskLower.includes("test") || taskLower.includes("spec")) {
    if (prompt.category === "testing") {
      return "Helps write comprehensive tests";
    }
  }

  if (taskLower.includes("bug") || taskLower.includes("fix") || taskLower.includes("debug")) {
    if (prompt.category === "debugging") {
      return "Systematic approach to finding and fixing issues";
    }
  }

  if (taskLower.includes("refactor") || taskLower.includes("clean") || taskLower.includes("improve code")) {
    if (prompt.category === "refactoring") {
      return "Helps restructure code for clarity and maintainability";
    }
  }

  // Generic tips based on matched fields
  if (matchedFields.includes("title")) {
    return `Directly relevant - matches "${prompt.title.toLowerCase()}"`;
  }

  if (matchedFields.includes("tags") && prompt.tags.length > 0) {
    return `Tagged with: ${prompt.tags.slice(0, 3).join(", ")}`;
  }

  if (matchedFields.includes("description")) {
    return prompt.description.length > 60
      ? prompt.description.slice(0, 57) + "..."
      : prompt.description;
  }

  // Default tip based on category
  const categoryTips: Record<string, string> = {
    ideation: "Helps generate creative solutions",
    documentation: "Maintains clear documentation",
    automation: "Streamlines repetitive tasks",
    refactoring: "Improves code structure",
    testing: "Ensures code quality",
    debugging: "Identifies and fixes issues",
    workflow: "Optimizes your process",
    communication: "Improves clarity in writing",
  };

  return categoryTips[prompt.category] || `${prompt.category} prompt`;
}

/**
 * Format a suggestion for JSON output
 */
function formatSuggestion(result: SearchResult, task: string, semantic?: boolean): SuggestOutput["suggestions"][0] {
  // Normalize relevance to 0-1 scale (semantic scores are 0-1, BM25 scores are ~0-5)
  const normalizedRelevance = Math.min(1, semantic ? result.score : result.score / 5);
  return {
    id: result.prompt.id,
    title: result.prompt.title,
    description: result.prompt.description,
    category: result.prompt.category,
    relevance: Math.round(normalizedRelevance * 100) / 100,
    matchedFields: result.matchedFields,
    tip: generateTip(result.prompt, task, result.matchedFields),
  };
}

File: crates/jfp/src/commands/suggest.rs
//! Suggest command implementation
//!
//! From EXISTING_JFP_STRUCTURE.md section 12 (suggest):
//! - Suggests prompts for a task description
//! - Uses FTS5 search as a simple relevance mechanism
//! - Semantic search option (not yet implemented)

use std::process::ExitCode;

use serde::Serialize;

use crate::registry::bundled_prompts;
use crate::storage::Database;

#[derive(Serialize)]
struct SuggestOutput {
    task: String,
    suggestions: Vec<Suggestion>,
    #[serde(skip_serializing_if = "Option::is_none")]
    semantic: Option<bool>,
}

#[derive(Serialize)]
struct Suggestion {
    id: String,
    title: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    description: Option<String>,
    relevance: f64,
    #[serde(skip_serializing_if = "Option::is_none")]
    reason: Option<String>,
}

pub fn run(task: &str, limit: usize, semantic: bool, use_json: bool) -> ExitCode {
    if semantic {
        if use_json {
            println!(r#"{{"error": "semantic_not_implemented", "message": "Semantic search not yet available"}}"#);
        } else {
            eprintln!("Semantic search not yet implemented. Using keyword search.");
        }
        // Continue with keyword search
    }

    // Open database
    let db = match Database::open() {
        Ok(db) => db,
        Err(e) => {
            if use_json {
                eprintln!(r#"{{"error": "database_error", "message": "{}"}}"#, e);
            } else {
                eprintln!("Error opening database: {}", e);
            }
            return ExitCode::FAILURE;
        }
    };

    // Seed if empty
    let count = db.prompt_count().unwrap_or(0);
    if count == 0 {
        let prompts = bundled_prompts();
        for prompt in &prompts {
            let _ = db.upsert_prompt(prompt);
        }
    }

    // Search for relevant prompts using FTS5
    let results = match db.search(task, limit) {
        Ok(r) => r,
        Err(e) => {
            if use_json {
                eprintln!(r#"{{"error": "search_error", "message": "{}"}}"#, e);
            } else {
                eprintln!("Error searching: {}", e);
            }
            return ExitCode::FAILURE;
        }
    };

    let suggestions: Vec<Suggestion> = results
        .into_iter()
        .map(|(prompt, score)| {
            // Generate a simple reason based on matching
            let reason = generate_reason(&prompt.title, &prompt.description, &prompt.tags, task);

            Suggestion {
                id: prompt.id,
                title: prompt.title,
                description: prompt.description,
                relevance: score,
                reason: Some(reason),
            }
        })
        .collect();

    if use_json {
        let output = SuggestOutput {
            task: task.to_string(),
            suggestions,
            semantic: if semantic { Some(false) } else { None },
        };
        match serde_json::to_string_pretty(&output) {
            Ok(json) => println!("{}", json),
            Err(e) => {
                eprintln!(r#"{{"error": "serialization_error", "message": "{}"}}"#, e);
                return ExitCode::FAILURE;
            }
        }
    } else {
        if suggestions.is_empty() {
            println!("No suggestions found for: {}", task);
            println!("\nTry different keywords or use 'jfp list' to browse all prompts.");
        } else {
            println!("Suggestions for: {}\n", task);

            for (i, s) in suggestions.iter().enumerate() {
                println!("{}. {} ({})", i + 1, s.title, s.id);
                if let Some(desc) = &s.description {
                    println!("   {}", desc);
                }
                if let Some(reason) = &s.reason {
                    println!("   Why: {}", reason);
                }
                println!();
            }

            println!("Use 'jfp show <id>' to see full details");
            println!("Use 'jfp copy <id>' to copy to clipboard");
        }
    }

    ExitCode::SUCCESS
}

/// Generate a simple reason for why a prompt was suggested
fn generate_reason(
    title: &str,
    description: &Option<String>,
    tags: &[String],
    task: &str,
) -> String {
    let task_lower = task.to_lowercase();
    let task_words: Vec<&str> = task_lower.split_whitespace().collect();

    let mut matches = Vec::new();

    // Check title
    let title_lower = title.to_lowercase();
    for word in &task_words {
        if title_lower.contains(word) && word.len() > 2 {
            matches.push(format!("title contains '{}'", word));
            break;
        }
    }

    // Check description
    if let Some(desc) = description {
        let desc_lower = desc.to_lowercase();
        for word in &task_words {
            if desc_lower.contains(word) && word.len() > 2 {
                matches.push(format!("description mentions '{}'", word));
                break;
            }
        }
    }

    // Check tags
    for tag in tags {
        let tag_lower = tag.to_lowercase();
        for word in &task_words {
            if tag_lower.contains(word) || word.contains(&tag_lower) {
                matches.push(format!("tagged as '{}'", tag));
                break;
            }
        }
    }

    if matches.is_empty() {
        "Related to your task keywords".to_string()
    } else {
        matches.join("; ")
    }
}
